xds/src/main/java/io/grpc/xds/internal/sds/DynamicSslContextProvider.java

```diff
@@ -0,0 +1,134 @@
+  @Nullable protected SslContext sslContext;
+
+  protected DynamicSslContextProvider(
+      BaseTlsContext tlsContext, CertificateValidationContext staticCertValidationContext) {
+    super(tlsContext);
+    this.staticCertificateValidationContext = staticCertValidationContext;
+  }
+
+  @Nullable
+  public SslContext getSslContext() {
+    return sslContext;
+  }
+
+  protected abstract CertificateValidationContext generateCertificateValidationContext();
+
+  /** Gets a server or client side SslContextBuilder. */
+  protected abstract SslContextBuilder getSslContextBuilder(
+          CertificateValidationContext certificateValidationContext)
+      throws CertificateException, IOException, CertStoreException;
+
+  // this gets called only when requested secrets are ready...
+  protected final void updateSslContext() {
+    try {
+      CertificateValidationContext localCertValidationContext =
+          generateCertificateValidationContext();
+      SslContextBuilder sslContextBuilder = getSslContextBuilder(localCertValidationContext);
+      CommonTlsContext commonTlsContext = getCommonTlsContext();
+      if (commonTlsContext != null && commonTlsContext.getAlpnProtocolsCount() > 0) {
+        List<String> alpnList = commonTlsContext.getAlpnProtocolsList();
+        ApplicationProtocolConfig apn =
+            new ApplicationProtocolConfig(
+                ApplicationProtocolConfig.Protocol.ALPN,
+                ApplicationProtocolConfig.SelectorFailureBehavior.NO_ADVERTISE,
+                ApplicationProtocolConfig.SelectedListenerFailureBehavior.ACCEPT,
+                alpnList);
+        sslContextBuilder.applicationProtocolConfig(apn);
+      }
+      SslContext sslContextCopy = sslContextBuilder.build();
+      sslContext = sslContextCopy;
+      makePendingCallbacks(sslContextCopy);
+    } catch (CertificateException | IOException | CertStoreException e) {
+      logger.log(Level.SEVERE, "exception in updateSslContext", e);
+    }
+  }
+
+  protected final void callPerformCallback(
+          Callback callback, final SslContext sslContextCopy) {
+    performCallback(
+        new SslContextGetter() {
+          @Override
+          public SslContext get() {
+            return sslContextCopy;
+          }
+        },
+        callback
+    );
+  }
+
+  @Override
+  public final void addCallback(Callback callback) {
+    checkNotNull(callback, "callback");
+    // if there is a computed sslContext just send it
+    SslContext sslContextCopy = sslContext;
+    if (sslContextCopy != null) {
+      callPerformCallback(callback, sslContextCopy);
+    } else {
+      synchronized (pendingCallbacks) {
+        pendingCallbacks.add(callback);
+      }
+    }
+  }
+
+  protected final void makePendingCallbacks(SslContext sslContextCopy) {
+    synchronized (pendingCallbacks) {
+      for (Callback callback : pendingCallbacks) {
+        callPerformCallback(callback, sslContextCopy);
```

Comment:
Why invoking callback here requires the lock while it doesn't in `addCallback`?

Reply:
`addCallback` has on line 110:
```
      synchronized (pendingCallbacks) {
        pendingCallbacks.add(callback);
      }
```
The `if` clause on line 107 does not require the lock

Reply:
> The if clause on line 107 does not require the lock

That is what I am concerning. You are holding the lock to perform the callback here. But in `addCallback` it does not. This means at least one case is wrong. If invoking the callback requires synchronization (that's very suspicious), the one in `addCallback` also needs to do so. If the callback does not need synchronization, a concurrent list here is sufficient.

Reply:
@voidzcy the lock is for the `pendingCallbacks` list only and not for the actual callback. When `addCallback` is called and there is an `sslContext` available then callback is called with that sslContext (as sslContextCopy). Otherwise `addCallback` and `makePendingCallbacks` have to synchronize on `pendingCallbacks`

Reply:
Then you don't need to hold the lock while invoking the callback, just make `pendingCallbacks` a concurrent list.

Reply:
That's a good point but I don't know if it's worth making the change/improvement. The list is unlikely to be longer than 1 or 2 elements (after the callbacks it is cleared) and the callbacks are into our own code (never user provided callbacks). Also by "concurrent list" did you mean `CopyOnWriteArrayList`?

Reply:
I was just being abstract of saying "concurrent list". But, realistically `CopyOnWriteArrayList` is an overkill. I suspect `ArrayBlockingQueue` would be enough.

Reply:
I addressed this comment by making a copy of the list for invoking the callback so there is no lock held when the callbacks are made. Using `ArrayBlockingQueue` wouldn't work here BTW.

Reply:
Why a blocking queue does not work? I don't see any get/insert-to-middle operations needed.

Reply:
The pendingCallbacks is only needed to hold the "pending callbacks" until an sslContext is built and appropriate locks should be in place to ensure there is no race between checking/updating of pendingCallbacks and sslContext (like the latest fix made).

I don't see the need for either blocking or queue aspect of the blocking queue.

Reply:
I mean using a blocking data structure is way better than your current approach (aka, acquire the lock, make a copy, etc). It's more efficient and elegant in terms of code maintainability.